compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2062,29 +2062,113 @@ impl<'tcx, T: HasTyCtxt<'tcx>> HasTyCtxt<'tcx> for LayoutCx<'tcx, T> {
     }
 }
 
+pub trait MaybeResult<T> {
+    type Error;
+
+    fn from(x: Result<T, Self::Error>) -> Self;
+    fn to_result(self) -> Result<T, Self::Error>;
+}
+
+impl<T> MaybeResult<T> for T {
+    type Error = !;
+
+    fn from(Ok(x): Result<T, Self::Error>) -> Self {
+        x
+    }
+    fn to_result(self) -> Result<T, Self::Error> {
+        Ok(self)
+    }
+}
+
+impl<T, E> MaybeResult<T> for Result<T, E> {
+    type Error = E;
+
+    fn from(x: Result<T, Self::Error>) -> Self {
+        x
+    }
+    fn to_result(self) -> Result<T, Self::Error> {
+        self
+    }
+}
```

Comment:
How does this work without specialization and not result in conflicting impls?

Reply:
As I understand it, trying to unify `Result<$0, _>` with `$0` (to check whether one of the `impl`s applies in the context of the other one) will fail because its sole "solution" is an infinitely nested `Result<Result<Result<...` - in the literature, I believe this is called an ["occurs check"](https://en.wikipedia.org/wiki/Occurs_check).

A crucial component of this, btw, is that `MaybeResult` is generic over the `Ok`/`T` type, instead of having it associated like the `Err`/`Error` type - if it was an associated type, it would indeed overlap.

This also shows up when implementing `From`: `impl<T> From<T> for T` doesn't involve specialization to avoid conflicting with your own different `impl`s.